### DISCUSSION TITLE
How to find the cert-label for smartcard eID with PKCS#11?

### DISCUSSION THREAD

**Discussion by Mato-Z** [MatthiasValvekens/pyHanko#201] [Q&A]
Hello, great program many thanks!  I have a problem finding the cert-id (and key-id) for my smartcard - Slovak eID with my qualified electronic signature. 
I used command: `pkcs11-tool --module /usr/lib/eID_klient/libpkcs11_x64.so -I -O`
Output is only:
```
Cryptoki version 2.20
Manufacturer     Ministerstvo vnutra SR
Library          eID SK PKCS#11 library (ver 4.0)
Using slot 0 with a present token (0x1)
```

Could you help me please with way what to find these data? Thanks a lot.

**Comment by Mato-Z** [comment]
If I use `pkcs11-tool --module /usr/lib/eID_klient/libpkcs11_x64.so -T`
Output is:
```
Available slots:
Slot 0 (0x1): Gemalto PC Twin Reader (E3FE96A2) 01 00; Sig_ZEP
  token label        : Sig_ZEP
  token manufacturer : Idemia Idenity & Security France
  token model        : Cosmo & CombICAO
  token flags        : login required, PIN pad present, token initialized, PIN initialized
  hardware version   : 9.2
  firmware version   : 2.2
  serial num         : ffffffff
  pin min/max        : 6/6
Slot 1 (0x2): Gemalto PC Twin Reader (E3FE96A2) 01 00; Sig_EP
  token label        : Sig_EP
  token manufacturer : Idemia Idenity & Security France
  token model        : Cosmo & CombICAO
  token flags        : login required, PIN pad present, token initialized, PIN initialized
  hardware version   : 9.2
  firmware version   : 2.2
  serial num         : ffffffff
  pin min/max        : 6/6
```
I have created simple config:
```
pkcs11-setups:
   test-setup:
       module-path: /usr/lib/eID_klient/libpkcs11_x64.so
       token-criteria:
           serial: ffffffff
           label: Sig_ZEP
       cert-label: Sig_ZEP
```
And I tested it: 
`pyhanko --verbose --config pyhanko.yml sign addsig --field Sig1 --timestamp-url MY_TSA_URL --use-pades pkcs11 --p11-setup test-setup test.pdf output.pdf`
Errors:
```
pkcs11.exceptions.PKCS11Error: No token matching criteria TokenCriteria(label='Sig_ZEP', serial=b'\xff\xff\xff\xff') found
Error: PKCS#11 error: [PKCS11Error] No token matching criteria TokenCriteria(label='Sig_ZEP', serial=b'\xff\xff\xff\xff') found
```
No matching token was found. :-( 

I think access to this data is also already PIN protected. Because government-provided applications will only display the slot selection itself after entering the access pin via own virtual keyboard. Is support for such smartcard also possible?

**Comment by Mato-Z** [comment]
I think that the main difference in access method is here: https://esignature.ec.europa.eu/efda/notification-tool/#/screen/browse/list/QSCD_SSCD, for example Slovak eID is declared like "QSCD" and Belgian eID is declared only like "SSCD". Do you think this is the reason for the different access to the data on the eID with PKCS#11?

**Reply by MatthiasValvekens** [reply]
Hi @Mato-Z 


Sorry for the delay in replying.

I don't think that list is up to date---if not, then the Belgian government is mis-issuing its citizen certs, since they claim that the private keys reside in QSCDs.

Either way, it shouldn't matter for the purposes of accessing the device over PKCS#11: the distinction between QSCDs and SSCDs is a policy/management issue, and has no bearing on the technical capabilities of the device.

Here's what I would try:

 - Did you check whether the token label contains spaces or other invisible characters/typographic oddities?
 - Did you try leaving off the serial number?
 - Failing that, did you try to manually enumerate the slots on the token using `get_token()` in `python-pkcs11`? `pyhanko` delegates pretty much all PKCS#11 handling to that library, so if something is not picked up correctly at that level, we at least know the problem is not in pyHanko itself.  (Note to self: I should probably implement a quick diagnostic function to enumerate all the tokens using `python-pkcs11` for this exact reason...)

Also be aware that the token label, key label and certificate label are all (a priori) different things that may or may not coincide for your specific device. They're respectively the label of the token managing your data, the label of the private key object on the token, and the label of the relevant certificate object on the token that corresponds to your private key. On some (but not all!) PKCS#11 implementations, enumerating certs/keys requires logging in to the token first. IIRC that can be done with some extra arguments to `pkcs11-tool`.

Long story short, the PKCS#11 integration in pyHanko tries very hard to not make too many assumptions about the structure of the token. The effect is that pretty much any PKCS#11 device can be used with it, but getting the configuration right might require some effort :).

**Reply by Mato-Z** [reply]
@MatthiasValvekens Many thanks for your reply. So:

1. It seems without invisible characters.
2. Yes, error is now: 
`pkcs11.exceptions.PKCS11Error: Could not find cert with label 'Sig_ZEP'. 
Error: Generic processing error.` So there is some progress. I tried empty label with `""` too... 
3. `get_token()` gives me: `Sig_ZEP`, so correctly.
Could you please give me some advice on how to find cert label? Thanks.

**Reply by MatthiasValvekens** [reply]
Try this: `pkcs11-tool --module /usr/lib/eID_klient/libpkcs11_x64.so --token-label Sig_ZEP -O`. That should list all objects in the token. Hopefully you'll find what you need there. If not, try again with the `-l` switch to do the same thing with a logged-in session.

**Reply by Mato-Z** [reply]
@MatthiasValvekens thanks, when I use `pkcs11-tool --module /usr/lib/eID_klient/libpkcs11_x64.so --token-label Sig_ZEP -l -O` (without -l output is empty) it shows me the virtual keyboard from my eID. I enter my correct PIN as usual and it ends with `Segmentation fault ` :-(

**Reply by Mato-Z** [reply]
@MatthiasValvekens I have some progress. I tried use PKCS11Admin on Windows OS. There I see that cert-label is `Signing Certificate`. After this I got this error: 
```
pkcs11.exceptions.NoSuchKey: No key matching {<Attribute.CLASS>: <ObjectClass.PRIVATE_KEY>, <Attribute.LABEL>: 'Signing Certificate'}
Error: Generic processing error.
```
In PKCS11Admin I see Key label: "Signing Key" and ID contains 9 hexa characters: `53 69 ... 5F 30` so I have added key-id (hexa) to config and I get this error now:  
```
  File "pkcs11/_pkcs11.pyx", line 322, in pkcs11._pkcs11.SearchIter.__next__
  File "pkcs11/_pkcs11.pyx", line 692, in pkcs11._pkcs11.Object._make
  File "pkcs11/_pkcs11.pyx", line 726, in pkcs11._pkcs11.Object.__getitem__
  File "pkcs11/_utils.pyx", line 11, in pkcs11._pkcs11.CK_BYTE_buffer
  File "stringsource", line 152, in View.MemoryView.array.__cinit__
OverflowError: Python int too large to convert to C ssize_t
Error: Generic processing error.
```
Any idea where is the problem please?

**Reply by MatthiasValvekens** [reply]
I'm not sure what's causing the overflow there, but from what you wrote, it looks like adding `--key-label "Signing Key"` to the pyHanko CLI should work :)

**Reply by Mato-Z** [reply]
@MatthiasValvekens thanks, but still the same error. I tried both (key-id and key-label) only key-label... At that point, another prompt for the second pin should appear, the so-called "signature pin"... Any other ideas please? :-(

**Reply by MatthiasValvekens** [reply]
Which error are you seeing with the `key-label`-only approach? Can you copy and paste the exact message? You could also try troubleshooting directly by enumerating all keys using python-pkcs11 (log in, and call `.get_objects({Attribute.CLASS: ObjectClass.PRIVATE_KEY})` on the resulting session object), but that requires a bit of work.

> At that point, another prompt for the second pin should appear, the so-called "signature pin"

Ah, this is definitely something that is not in scope of the PKCS#11 interface standard, so it's something private to the implementation that you're using. Are you running this code on a headless machine? Is it perhaps trying to use `pinentry` without a proper pinentry client configured? If you're not seeing a pyHanko error and the process still fails for some reason, that could be something worth investigating.

**Reply by Mato-Z** [reply]
@MatthiasValvekens Thanks. Slovak eID uses at least two PINs. First - access to token and second for the signature itself. And I think there is one more PIN for some blocking/unblocking. Belgian eID uses only one PIN? 
key-label-only approach output is:
```
pyhanko --verbose --config pyhanko.yml sign addsig --field Sig1 --timestamp-url TSA_URL --use-pades pkcs11 --p11-setup test-setup test.pdf output.pdf
2023-01-06 22:57:08,936 - root - DEBUG - Running with --verbose
2023-01-06 22:57:08,937 - root - DEBUG - Finished reading configuration from <_io.TextIOWrapper name='pyhanko.yml' mode='r' encoding='UTF-8'>.
PKCS#11 user PIN: 
2023-01-06 22:57:13,611 - pyhanko.cli - ERROR - Generic processing error.
Traceback (most recent call last):
  File "/home/mato/.local/lib/python3.10/site-packages/pyhanko/cli.py", line 98, in pyhanko_exception_manager
    yield
  File "/home/mato/.local/lib/python3.10/site-packages/pyhanko/cli.py", line 1115, in _sign_pkcs11
    generic_sign_pdf(
  File "/home/mato/.local/lib/python3.10/site-packages/pyhanko/cli.py", line 922, in generic_sign_pdf
    result = signers.PdfSigner(
  File "/home/mato/.local/lib/python3.10/site-packages/pyhanko/sign/signers/pdf_signer.py", line 917, in __init__
    mech = self.signer.get_signature_mechanism(None)
  File "/home/mato/.local/lib/python3.10/site-packages/pyhanko/sign/signers/pdf_cms.py", line 342, in get_signature_mechanism
    if self.signing_cert is None:
  File "/home/mato/.local/lib/python3.10/site-packages/pyhanko/sign/pkcs11.py", line 557, in signing_cert
    self._load_objects()
  File "/home/mato/.local/lib/python3.10/site-packages/pyhanko/sign/pkcs11.py", line 673, in _load_objects
    kh = self.pkcs11_session.get_key(
  File "/home/mato/.local/lib/python3.10/site-packages/pkcs11/types.py", line 324, in get_key
    key = next(iterator)
  File "pkcs11/_pkcs11.pyx", line 322, in pkcs11._pkcs11.SearchIter.__next__
  File "pkcs11/_pkcs11.pyx", line 692, in pkcs11._pkcs11.Object._make
  File "pkcs11/_pkcs11.pyx", line 726, in pkcs11._pkcs11.Object.__getitem__
  File "pkcs11/_utils.pyx", line 11, in pkcs11._pkcs11.CK_BYTE_buffer
  File "stringsource", line 152, in View.MemoryView.array.__cinit__
OverflowError: Python int too large to convert to C ssize_t
Error: Generic processing error.
```
This eID works for example with Java https://docs.oracle.com/javase/8/docs/technotes/guides/security/p11guide.html I can provide you example of implementation...

**Reply by MatthiasValvekens** [reply]
Hi @Mato-Z, sorry about the delay (dealing with a newborn at home right now :) ).

This is starting to look like a bug in `python-pkcs11` or possibly an issue with the Slovak eID PKCS#11 implementation that causes `python-pkcs11` to choke on it. I think the best way to deal with this would be to try to reproduce the error in isolation (=> make a `python-pkcs11` call to the token to retrieve the key in the way pyHanko does it), and to use that to file a bug report over here: https://github.com/danni/python-pkcs11.

**Reply by Mato-Z** [reply]
@MatthiasValvekens it is ok, good luck :-)
python-pkcs11 uses OpenSC for access to smatcards right?
I found that the problem is probably here https://github.com/OpenSC/OpenSC/issues/208#issuecomment-1319429080
Slovak eID implementation is officially unfortunately closed, but here is good technical analysis how it works https://github.com/OpenSC/OpenSC/issues/208#issuecomment-1319429080
And today I found information that some "good guy" is trying to add Slovak eID support to OpenSC https://github.com/jurajsarinay/OpenSC/tree/skeid, but for now only for old version of Slovak eID with CardOS 5.4. Currently this version is for qualified signature unsupported. I hope that new version will be supported too in short time. So with new version of OpenSC (with support for Slovak eID) it can work (I hope)?

**Reply by MatthiasValvekens** [reply]
> python-pkcs11 uses OpenSC for access to smatcards right?

`python-pkcs11` is a completely generic PKCS#11 implementation as far as I'm aware; it doesn't depend on OpenSC as such, and it doesn't "know" whether or not it's even talking to a smartcard at all. From your comments, it also appears as if the PKCS#11 module implementation you use is also not OpenSC.

Having said that, it's not a totally dead end:

- OpenSC typically operates at a lower level on the stack => it can do more than just PKCS#11 operations, and IIRC OpenSC ships with a generic PKCS#11 module implementation. If OpenSC were to implement support for Slovakian ID cards, you could try replacing the "official" PKCS#11 module (`/usr/lib/eID_klient/libpkcs11_x64.so`) with the generic OpenSC one.
- It's IMHO unlikely but not impossible that the problem you're seeing in `python-pkcs11` is also the reason why the smartcard currently doesn't work with OpenSC (EDIT: especially the bits about invalid PKCS#15 structures on the card).

I would still try reproing the problem using `python-pkcs11` exclusively, and raise an issue over there once you have a reliable, minimal code sample to demonstrate---the library author probably won't have access to a device like yours, so it's important to nail down the exact circumstances that cause the problematic behaviour.

**Reply by Mato-Z** [reply]
@MatthiasValvekens many thanks for the explanation, that would also explain correct running with Java, which uses "native module" `/usr/lib/eID_klient/libpkcs11_x64.so `
>  think the best way to deal with this would be to try to reproduce the error in isolation (=> make a `python-pkcs11` call to the token to retrieve the key in the way pyHanko does it), and to use that to file a bug report over here: https://github.com/danni/python-pkcs11.

Sure yes, but I'm afraid my knowledge in this area is insufficient to describe the problem in detail and demonstrate it outside of pyHanko...
